File: 3180/main.go
package main

import (
	"fmt"
	"sort"
)

func f(n int) (r int) {
	defer func() {
		r += n
		recover()
	}()

	var f func()
	defer f()
	f = func() {
		r += 2
	}
	return n + 1
}

func main() {
	fmt.Println(f(3))	// 7
	fmt.Println(maxTotalReward([]int{1,1,3,3}))	// 4
	s := []string{"a", "b", "c", "d"}
	defer fmt.Println(s)
	defer func() {
		fmt.Println(s[:1], len(s[:1]), cap(s[:1]))
		_ = append(s[:1], "x", "y")
	}()
}

func maxTotalReward(rewardValues []int) int {
	sort.Ints(rewardValues)
	m := rewardValues[len(rewardValues)-1]
	dp := make([]int, 2 * m)
	dp[0] = 1
	for _, x := range rewardValues {
		for k := 2 * x - 1; k >= x; k-- {
			if dp[k-x] == 1 {
				dp[k] = 1
			}
		}
	}
	x := 0
	for i := 0; i < len(dp); i++ {
		if dp[i] == 1 {
			x = i
		}
	}

	return x
}

File: 3180/main.rs
struct Solution {}

impl Solution {
    pub fn max_total_reward(reward_values: Vec<i32>) ->i32 {
        let mut reward_values = reward_values.clone();
        reward_values.sort();
        let m = reward_values[reward_values.len()-1] as usize;
        let mut dp = vec![0; 2 * m];
        dp[0] = 1;
        for &x in &reward_values {
            let x = x as usize;
            for k in (x..2 * x).rev() {
                if dp[k-x] == 1 {
                    dp[k] = 1
                }
            }
        }
        let mut res = 0;
        for (i, &value) in dp.iter().enumerate() {
            if value == 1 {
                res = i;
            }
        }
        res as i32
    }
}

fn main() {
    println!("{}", Solution::max_total_reward(vec![1,1,3,3]));
}